### PR TITLE
gpu fix

### PR DESCRIPTION
**Pull request by vitskvara**
Trying to solve https://github.com/aicenter/ConditionalDists.jl/issues/11.

### CODE REVIEW

File: src/cmean_gaussian.jl [hunk 1]
```diff
@@ -65,7 +65,7 @@ mean_var(p::CMeanGaussian, z::AbstractArray) = (mean(p, z), variance(p, z))
 function Flux.functor(p::CMeanGaussian{V,S,M}) where {V,S,M}
     fs = fieldnames(typeof(p))
     nt = (; (name=>getfield(p, name) for name in fs)...)
-    nt, y -> CMeanGaussian{V,S,M}(y...)
+    nt, y -> CMeanGaussian{V,S,typeof(gpu(p.mapping))}(y...)
```
Comment (vitskvara) [inline src/cmean_gaussian.jl]: This only fixes the mapping conversion, however the `σ` field is still not converted.

Comment (vitskvara) [reply src/cmean_gaussian.jl]: Also, this is probably going to break if one has `CuArrays` loaded, the model is on cpu and functor is not called in the gpu conversion process - what are the places where `functor` is called?

**Comment by nmheim** [comment]
`functor` is used by e.g. `trainable` and `gpu`. it recursively tries to adapt all Arrays to CuArrays. As far as I can see it constructs the type signature (with the cpu arrays in the mapping) before moving the mapping to the gpu... I am not sure if this is easily solvable with the type signature we have right now...

**Comment by nmheim** [comment]
I cannot test on GPU atm, but does this work?
```julia
function Flux.functor(p::CMeanGaussian{V}) where V
    fs = fieldnames(typeof(p))
    nt = (; (name=>getfield(p, name) for name in fs)...)
    nt, y -> CMeanGaussian{V}(y...)
end
```

**Comment by nmheim** [comment]
putting a `gpu` call in `functor` is not a good idea, because `gpu` itsself calls `functor` (through `fmap` and `fmap1`). have you tried if the snippet I posted above works on a GPU?

**Comment by vitskvara** [comment]
It helps with `CMeanVarGaussian`, but not with `CMeanGaussian`.

**Comment by vitskvara** [comment]
Now it fails with
```julia
julia> p  = CMeanGaussian{DiagVar}(mapping, var) |> gpu
ERROR: MethodError: no method matching CMeanGaussian{DiagVar,M,S} where S<:AbstractArray where M(::Dense{typeof(identity),CuArray{Float32,2,Nothing},CuArray{Float32,1,Nothing}}, ::CuArray{Float32,1,Nothing}, ::Int64, ::Dict{Symbol,Bool})
Closest candidates are:
  CMeanGaussian{DiagVar,M,S} where S<:AbstractArray where M(::M, ::Any, ::Int64) where {V, M} at /home/vit/.julia/dev/ConditionalDists/src/cmean_gaussian.jl:40
  CMeanGaussian{DiagVar,M,S} where S<:AbstractArray where M(::Any, ::Any) at /home/vit/.julia/dev/ConditionalDists/src/cmean_gaussian.jl:46
Stacktrace:
 [1] #14 at /home/vit/.julia/dev/ConditionalDists/src/cmean_gaussian.jl:68 [inlined]
 [2] fmap1(::Function, ::CMeanGaussian{DiagVar,Dense{typeof(identity),Array{Float32,2},Array{Float32,1}},Array{Float32,1}}) at /home/vit/.julia/packages/Flux/oObnA/src/functor.jl:32
 [3] #fmap#41(::IdDict{Any,Any}, ::typeof(fmap), ::typeof(cu), ::CMeanGaussian{DiagVar,Dense{typeof(identity),Array{Float32,2},Array{Float32,1}},Array{Float32,1}}) at /home/vit/.julia/packages/Flux/oObnA/src/functor.jl:37
 [4] fmap at /home/vit/.julia/packages/Flux/oObnA/src/functor.jl:36 [inlined]
```

**Comment by vitskvara** [comment]
I am really confused with this. When debugging, it seems that the overloaded `functor` is called twice before some default one is called which tries to construct the `Dense` layer but with the original `CMeanGaussian` arguments, which of course does not make any sense.

**Comment by nmheim** [comment]
`gpu` tried to call `CMeanGaussian{V}(m,s,xlength,d)` at some point, but only `CMeanGaussian{V,M,S}(m,s,xlength,d)` (with the types `M` and `S`) was defined.

**Pull request merged**